### DISCUSSION TITLE
Planet X3 game on ELKS

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1617] [Ideas]
I think there is a good chance that Planet X3 can be built for ELKS. 

https://github.com/planet-x3/px3_ose

Build proces: https://github.com/planet-x3/px3_ose/wiki/Build-process

The game is written in x86 assembly using YASM.
YASM cas support the following formats: https://www.tortall.net/projects/yasm/manual/html/manual.html#running-objfmt
So YASM needs to support the elf format of ELKS or the final elf executable must be converted to ELKS' elf format.  Does it sound right?

**Comment by ghaerr** [comment]
ELKS doesn't actually support an ELF format, but the `ia16-elf-gcc` toolchain produces 16-bit output in an ELF format, which is then translated to the ELKS/MINIX a.out format using `elf2elks`. It is unclear whether the 32-bit ELF formats that YASM can use are compatible with `elf2elks`, that would have to be tested.

With regards to porting the ASM files, a quick looks shows that it hooks the hardware timer interrupt and reprograms the 8253/8254 programmable interval timer, which is not allowed by ELKS, so that would need rewriting. In addition, the sound portion appears to handle output by direct manipulation, which may be risky with regards to ELKS kernel operation. All the disk I/O would need to be rewritten to call ELKS rather than the current int 21h DOS calls.

**Comment by toncho11** [comment]
Thank you @ghaerr.

The advantage here is that we have the source code of a big DOS game that compiles using Linux tools. 
Also there are youtube videos that explain how the game is made: https://www.youtube.com/watch?v=BdhvIfmRtvQ&ab_channel=The8-BitGuy
But as always it is still an investment in time.
